### DISCUSSION TITLE
EBSDMeshGenerator/Reader + Distributed Mesh Workflow

### DISCUSSION THREAD

**Discussion by gka80** [idaholab/moose#17442] [Q&A General] [answered]
Hello-

I'm working on a modification to the EBSDMeshGenerator/Reader workflow to take in data of a different format (e.g. particle geometry developed in a different program). For this application, I have the workflow operating quite nicely. The issue that I'm running into currently is that using a replicated mesh, the memory requirements are intensive. Some notes:

1. The goal is to run this on a cluster. Each node has access to 28 cores and 100 GB (useable) memory. Simple tests indicate that (100/28) ~3.5 GB/core is "too low." The run shown in the file attachment was terminated at ~10 GB/core which means that I would be using less than 50% of the node core capacity which is too inefficient (as I don't have access to many nodes to begin with and am restricted to 10 hour increments). Tested on a PC with 64 GB memory.

2. The "coarse" (test) mesh is ~500,000 elements (HEX8). Small. I won't be running mesh adaptivity, but I would like to refine this mesh further for the final run.

3. The number of non-linear variables is ~20. Small but not inconsequential.

4. The number of variables in the auxiliary system isn't what I would consider "excessive" (although the number of DOF seems rather high [possibly due to the tensor mobility]).

5. I've read and understand the mesh splitting documentation [(Mesh Splitting)](https://mooseframework.inl.gov/syntax/Mesh/splitting.html) and have successfully used it on a few test problems. However, while I have an idea as to why the current workflow doesn't work as distributed, I'm unable to pinpoint "what" in the MOOSE framework needs to be implemented to make it work (I know it has to do with libMesh).

6. bash output attached as an example.

7. EDIT: Also, SMP preconditioning using ASM with "sub" preconditioning type ilu with 3 factor levels and an overlap of 1. I've tried other preconditioning options but found this combination to work the best on smaller test problems. Admittedly, this area is a bit new to me (I've read the PETSc documentation but didn't understand _everything_; have yet to make a second pass). AMG-based methods performed poorly but I've read that that could be because I'm using a split CH formulation (could also be because I'm a beginner).

Does anyone:

1. Have any recommendations on a "work around" to be able to split the problem and run the model as distributed? 
2. Have any examples (or is willing to point me in the right direction) so that I can modify the EBSDMeshGenerator/Reader workflow so that it can be implemented as distributed?

I know my problem is small, however I'm in personally uncharted territory. Not opposed to learning, just hoping someone can throw me a bone.

Thank you!

-Garrett
![Screen Shot 2021-03-28 at 12 38 05](https://user-images.githubusercontent.com/30502230/112765558-8a33a580-8fc2-11eb-8c40-c2e79e7c0637.png)

**Comment by GiudGiud** [comment, marked as answer]
git blame says @dschwen is the expert here

**Reply by dschwen** [reply]
Ugh, totally missed this thread. I'll take a look tomorrow!

**Reply by dschwen** [reply]
Ok, it would be relatively straight forward to create a `DistributedEBSDMeshGenerator` (which inherits from `DistributedRectilinearMeshGenerator` rather than `GeneratedMeshGenerator`). I can do that in the next couple of days.

**Reply by gka80** [reply]
@dschwen Great! Thank you!

**Reply by dschwen** [reply]
Check out https://github.com/idaholab/moose/pull/19155 I'd welcome any feedback on it.

**Reply by gka80** [reply]
@dschwen Thank you for turning this around so quickly. I'll give this a try by COB Friday (or wait until the potential implementation with sub-mesh generation?) and offer comments (if any).

**Reply by dschwen** [reply]
SubGenerator version is now merged into MOOSE. To take advantage of it just add `parallel_type = DISTRIBUTED` in the `[Mesh]` block.

**Reply by gka80** [reply]
@dschwen @GiudGiud 

Thank you. It looks like this solution will work great. I'll chime in if I have any other questions.

**Comment by gka80** [comment]
@dschwen (or anyone), any hints, sample codes, or links/advice on "where to dig"? I can provide more information, if necessary.

**Reply by GiudGiud** [reply]
Hey @gka80 are you still working on this?
We can hit up more people if so

**Reply by gka80** [reply]
Hey @GiudGiud,

I was sidetracked with another project. However, this problem is still near the top of my priority list. I anticipate on getting back to it in the next few weeks. 

If someone has capacity to look into the issue, that'd be great. I'm still not opposed to coding up a solution myself if someone was able to provide some guidance. Whatever is easiest.

**Reply by GiudGiud** [reply]
@SudiptaBiswas  @jain651 @laagesen  any of you can offer guidance on that one?
